### PR TITLE
Updated AWS Kubernetes yaml.

### PR DESCRIPTION
**Pull request by ravichaturvedi**
Updated AWS Kubernetes yaml to use StatefulSet per stripe and initContainers for dependency across deployment. Also using persistent volume claim template so storage will be automatically provisioned using the default storage class registered in the Kubernetes cluster.

Instructions to deploy:
- Add registry credentials
kubectl create secret docker-registry regcred --docker-server=https://index.docker.io/v1/ --docker-username=<your-name> --docker-password=<your-pword> --docker-email=<your-email>

- Add config maps
kubectl create configmap tc-configs --from-file=./kubernetes/aws-kops/stripe1.xml --from-file=./kubernetes/aws-kops/stripe2.xml
    kubectl create configmap license --from-file=./license.xml

- Deploy terracotta
kubectl apply -f kubernetes/aws-kops/k8s.yaml

TBD Items:
- Update README
- Add Readiness Check

### CODE REVIEW

File: kubernetes/aws-kops/k8s.yaml [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+          command: ["bin/start-tc-server.sh"]
+          args: ["-f", "/configs/stripe1.xml", "-n", "$(POD_NAME)"]
+          imagePullPolicy: Always
+          env:
+            - name: ACCEPT_EULA
+              value: "Y"
+            - name: POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
```
Comment (anthonydahanne) [inline kubernetes/aws-kops/k8s.yaml]: pretty cool

File: kubernetes/aws-kops/k8s.yaml [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: tmc-internal
+  labels:
+    app: tmc
+spec:
+  ports:
+  - name: http
+    port: 9889
```
Comment (anthonydahanne) [inline kubernetes/aws-kops/k8s.yaml]: nope, no port 9889 in tcdb

File: kubernetes/aws-kops/k8s.yaml [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: tmc
+spec:
+  selector:
+    matchLabels:
+      app: tmc
+  serviceName: tmc-internal
```
Comment (anthonydahanne) [inline kubernetes/aws-kops/k8s.yaml]: hum... super weird... even more since tmc-internal exposes a non used port, 9889

**Comment by anthonydahanne** [comment]
@ravichaturvedi I've just deployed it locally (kops 1.9) and while it mostly worked, here are some remarks : 

* we should leverage node labels by default (readme appendix A); during my deployment some terracotta servers were deployed on clientnodes that also run some clients

* backup volumes were mounted to EBS (the default)- not great (an EFS volume with ReadWriteMany would make more sense for backups), I wonder why you discarded the existing PVC; for backups actually just 1 single PVC / PV should exist - maybe just keep things like they were : document backup but keep it commented / not by default 

* see my remarks with tmc-internal, not sure what's it's supposed to do here

* init containers for clients : well, it's kinda doing the work twice , since the clients entrypoints currently also run the cluster tool - that's ok for now but eventually we'll need to remove the cluster tool form the entry point and keep the init containers

* I'm not sure the tmc needs init container with cluster tool, since after all, it does not automatically create a connection to the cluster; so it does not need to wait and could be started even if the cluster is not yet ready

all in all, good  addition, thank you !

File: kubernetes/aws-kops/k8s.yaml (hunk 2 above)
Comment (ravichaturvedi) [reply kubernetes/aws-kops/k8s.yaml]: yup. missed to update the port. The deployment is not having any issue as headless service do not route the traffic.

File: kubernetes/aws-kops/k8s.yaml [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+  ports:
+  - name: tmc-port
+    port: 9480
+  type: LoadBalancer
+  selector:
+    app: tmc
+
+---
+
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (ravichaturvedi) [reply kubernetes/aws-kops/k8s.yaml]: Needed by the tmc stateful set, as StatefulSet need a supporting headless service. 
The other tmc service is actually accepting traffic from the load balancer.

**Comment by ravichaturvedi** [comment]
@anthonydahanne, please find the comments inline:

we should leverage node labels by default (readme appendix A); during my deployment some terracotta servers were deployed on clientnodes that also run some clients
_[ravi]: Was thinking to having a smaller cluster. will add it back._

backup volumes were mounted to EBS (the default)- not great (an EFS volume with ReadWriteMany would make more sense for backups), I wonder why you discarded the existing PVC; for backups actually just 1 single PVC / PV should exist - maybe just keep things like they were : document backup but keep it commented / not by default
[ravi] yes, ReadWriteMany would work fine and cost low storge. Adding the backup pv file back.

see my remarks with tmc-internal, not sure what's it's supposed to do here.
_[ravi] It is required by the tmc stateful set, as StatefulSet need a supporting headless service. Using stateful set as it would provision volume atomatically based on template. The other tmc service is actually accepting traffic from the load balancer._

init containers for clients : well, it's kinda doing the work twice , since the clients entrypoints currently also run the cluster tool - that's ok for now but eventually we'll need to remove the cluster tool form the entry point and keep the init containers
_[ravi] I ran into an issue where the client found the cluster is configured but license is not installed so entity creation was failing. And as the client image not existing/retrying, it was getting stuck.
I believe it's a race condition as cluster tool configure command runs a two step process of `configuring` and `installing license`.
As there is no command in cluster-tool to check the license status. I kindof using the status command in init-container to get some time before the container is actually started._

I'm not sure the tmc needs init container with cluster tool, since after all, it does not automatically create a connection to the cluster; so it does not need to wait and could be started even if the cluster is not yet ready
_[ravi] Make sense, will be removing the init-container. I was thinking of having a total dependency on the terracotta cluster configuration, as otherwise, TMC would fail connecting to the cluster, if stripe needs to load good amount of data from the file system._

**Comment by anthonydahanne** [comment]
ok, with kops 1.9.1, that worked perfectly !
I was able to trigger a fail over under a 50 pounding clients load, that worked like a charm !
Thank you very much @ravichaturvedi  !

**Pull request merged**